### PR TITLE
Default labels for app CRs

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#11656

```yaml
#nginx-ic.yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  name: nginx-ingress-controller
  namespace: 3zjhk
spec:
  catalog: default
  kubeConfig:
    inCluster: false
  name: nginx-ingress-controller-app
  namespace: kube-system
  version: 1.10.0
```

```bash
k apply -f nginx-ic.yaml
k -n 3zjhk get app nginx-ingress-controller -o yaml
```

```yaml
apiVersion: application.giantswarm.io/v1alpha1
kind: App
metadata:
  labels:
    app-operator.giantswarm.io/version: 2.3.5
    app.kubernetes.io/name: nginx-ingress-controller-app
  name: nginx-ingress-controller
  namespace: 3zjhk
spec:
  catalog: default
  config:
    configMap:
      name: ingress-controller-values
      namespace: 3zjhk
  kubeConfig:
    context:
      name: 3zjhk
    inCluster: false
    secret:
      name: 3zjhk-kubeconfig
      namespace: 3zjhk
  name: nginx-ingress-controller-app
  namespace: kube-system
  version: 1.10.0
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,8 +3,8 @@ module github.com/giantswarm/app-admission-controller
 go 1.15
 
 require (
-	github.com/giantswarm/apiextensions/v3 v3.7.0
-	github.com/giantswarm/app/v3 v3.2.0
+	github.com/giantswarm/apiextensions/v3 v3.7.1-0.20201111144349-4ed00f3702ef
+	github.com/giantswarm/app/v3 v3.2.1-0.20201111144832-2ec3df0b5723
```
Comment (rossf7) [inline go.mod:7]: TODO: Needs https://github.com/giantswarm/app/pull/39

File: pkg/app/mutate_app.go [hunk 2]
```diff
@@ -155,3 +168,46 @@ func (m *Mutator) mutateKubeConfig(ctx context.Context, app v1alpha1.App) ([]mut
 
 	return result, nil
 }
+
+func (m *Mutator) mutateLabels(ctx context.Context, app v1alpha1.App) ([]mutator.PatchOperation, error) {
+	var result []mutator.PatchOperation
+
+	// Set app label if there is no app label present.
+	if key.AppKubernetesNameLabel(app) == "" && key.AppLabel(app) == "" {
+		result = append(result, mutator.PatchAdd(fmt.Sprintf("/metadata/labels/%s", replaceToEscape(label.AppKubernetesName)), key.AppName(app)))
```
Comment (rossf7) [inline pkg/app/mutate_app.go:177]: Setting the `app.kubernetes.io/name` label for `kgs validate apps`. @oponder Can you confirm this is the correct label?

File: pkg/app/mutate_app.go [hunk 3]
```diff
@@ -155,3 +168,46 @@ func (m *Mutator) mutateKubeConfig(ctx context.Context, app v1alpha1.App) ([]mut
...
+	}
+
+	// Set version label to be the same as the chart-operator app CR. This
+	// is the version we need and means we don't need to check for a cluster CR.
```
Comment (rossf7) [inline pkg/app/mutate_app.go:181]: Like the comment says this is so the admission controller is not coupled with cluster or release CRs.

Once we run an app-operator per cluster I think we can remove the version label. As we will watch all CRs in the tenant cluster namespace in the control plane cluster.

**Review by rossf7 — COMMENTED** [review]
Tested on `ginger`.

File: pkg/app/mutate_app.go (hunk 2 above)
Comment (oponder) [reply pkg/app/mutate_app.go:177]: Looks correct to me!

Comment (rossf7) [reply pkg/app/mutate_app.go:177]: Thanks, it sets the label if there is no `app` or `app.kubernetes.io/name` for compatibility with existing app CRs managed by cluster-operator.

**Review by oponder — APPROVED** [review]
LGTM!

**Review by tomahawk28 — APPROVED** [review]
LGTM

**Comment by rossf7** [comment]
Tested on `giraffe`.

**Pull request merged**